Automatic language detection

arnavkapoor opened the pull request:
Added a simple count-based method to check the best language fit.

arnavkapoor commented on number_parser/parser.py:
```diff
@@ -200,8 +200,28 @@ def _apply_cardinal_conversion(token, lang_data):  # Currently only for English
+        tokens = _tokenize(input_string, language)
+        normalized_tokens = _normalize_tokens(tokens)
+        valid_list = [_is_number_token(token, lang_data) for token in normalized_tokens]
+        cnt_valid_words = sum(word is not None for word in valid_list)
```
The return value for _is_number_token isn't a boolean it's actually either ```None``` or the ```cardinal form ``` of the number.

noviluni replied:
I know that, but `word is not None` it's a boolean.

What you are doing here: `sum(word is not None for word in valid_list)` in case of "'I have three oranges and two apples" would be something like this:

`valid_list = [None, None, 3, None, None, 2, None]`

and then and you are summing up:  `sum([False, False, True, False, False, True, False])`, so what you are doing is:

`False+False+True+False+False+True+False` (which is `2`).

Do you understand the point now?

arnavkapoor replied:
Ah yes understood now. Sorry for the confusion. I have updated the code and now it uses ```list.count(True)```. I did look into the best sort of method and came across this question. https://stackoverflow.com/questions/12765833/counting-the-number-of-true-booleans-in-a-python-list which does seem to say that ```bool``` is a subclass of integer.

noviluni replied:
As you already know, one thing is what you can do and works and the other thing is what it's readable and logical :slightly_smiling_face: 

When coding you usually spend more time reading than writing, so this will be read a lot of times. Summing booleans is something weird and hard to understand, so it could confuse people when reading the code (and when I say "people" I also mean yourself in the future :smile: ).

Even if there is a more efficient way to do something, people try to be more explicit, and if the performance it's really important, we can proceed with that strange approach by adding a comment to clarify the intentions and motivations.

In this case, as you can read in the second answer from StackOverflow, using `count` "is more efficient than `sum`, as well as being more explicit about the intent", so it's just better.

Anyway, good job googling for solutions to what you are trying to do :100:

arnavkapoor replied:
Yes very true @noviluni , Thanks for a lot for all the feedback throughout the program. Till now most of my projects had always been focused on getting the thing to work only. So it has been a great learning experience to emphasize more on structured , easily readable codes.

arnavkapoor commented on number_parser/parser.py:
```diff
@@ -200,8 +200,28 @@ def _apply_cardinal_conversion(token, lang_data):  # Currently only for English
+        lang_data = LanguageData(language)
+        tokens = _tokenize(input_string, language)
+        normalized_tokens = _normalize_tokens(tokens)
+        valid_list = [_is_number_token(token, lang_data) for token in normalized_tokens]
```
Sure, I have made some changes to accommodate this.

noviluni commented on number_parser/__init__.py:
```diff
@@ -1 +1 @@
-from number_parser.parser import parse, parse_number, parse_ordinal
+from number_parser.parser import parse, parse_number, parse_ordinal, _valid_tokens_by_language
```
As this is a private function  I prefer no to add it here. You can access from the tests directly with `from number_parser.parser import _valid_tokens_by_language`

arnavkapoor commented:
Hi @noviluni  sorry for no updates over the weekend was occupied with something else , I have updated the code with the requested changes. I have also added some unit tests too for the language detection part.

noviluni commented on number_parser/parser.py:
```diff
@@ -181,6 +181,12 @@ def _is_number_token(token, lang_data):
     return _is_cardinal_token(token, lang_data)
 
 
+def _is_skip_token(token, lang_data):
+    if token in lang_data.skip_tokens:
+        return True
+    return False
```
```suggestion
def _is_skip_token(token, lang_data):
    return token in lang_data.skip_tokens
```
